src/goto-programs/goto_trace.cpp

```diff
@@ -348,6 +348,136 @@ bool is_index_member_symbol(const exprt &src)
     return false;
 }
 
+/// \brief show a compact variant of the goto trace on the console
+/// \param out the output stream
+/// \param ns the namespace
+/// \param goto_trace the trace to be shown
+/// \param options any options, e.g., numerical representation
+void show_compact_goto_trace(
+  messaget::mstreamt &out,
+  const namespacet &ns,
+  const goto_tracet &goto_trace,
+  const trace_optionst &options)
+{
+  std::size_t function_depth = 0;
+
+  for(const auto &step : goto_trace.steps)
+  {
+    if(step.is_function_call())
+      function_depth++;
+    else if(step.is_function_return())
+      function_depth--;
+
+    // hide the hidden ones
+    if(step.hidden)
+      continue;
+
+    switch(step.type)
+    {
+    case goto_trace_stept::typet::ASSERT:
+      if(!step.cond_value)
+      {
+        out << '\n';
+        out << messaget::red << "Violated property:" << messaget::reset << '\n';
+        if(!step.pc->source_location.is_nil())
+          out << "  " << state_location(step, ns) << '\n';
+
+        out << "  " << messaget::red << step.comment << messaget::reset << '\n';
+
+        if(step.pc->is_assert())
+          out << "  " << from_expr(ns, step.function, step.pc->guard) << '\n';
+
+        out << '\n';
+      }
+      break;
+
+    case goto_trace_stept::typet::ASSIGNMENT:
+      if(
+        step.assignment_type ==
+        goto_trace_stept::assignment_typet::ACTUAL_PARAMETER)
+      {
+        break;
+      }
+
+      out << "  ";
+
+      if(!step.pc->source_location.get_line().empty())
+      {
+        out << messaget::faint << step.pc->source_location.get_line() << ':'
+            << messaget::reset << ' ';
+      }
+
+      trace_value(
+        out,
+        ns,
+        step.get_lhs_object(),
+        step.full_lhs,
+        step.full_lhs_value,
+        options);
+      break;
+
+    case goto_trace_stept::typet::FUNCTION_CALL:
+      // downwards arrow
+      out << "\n" << messaget::faint << u8"\u21b3" << messaget::reset << ' ';
+      if(!step.pc->source_location.get_file().empty())
+      {
+        out << messaget::faint << step.pc->source_location.get_file();
+
+        if(!step.pc->source_location.get_line().empty())
+        {
+          out << messaget::faint << ':' << step.pc->source_location.get_line();
+        }
+
+        out << messaget::reset << ' ';
+      }
+
+      {
+        // show the display name
+        const auto &f_symbol = ns.lookup(step.called_function);
+        out << f_symbol.display_name();
+      }
+
+      out << '(';
+
+      {
+        bool first = true;
```

Comment:
:bulb: 
The pattern is very simple, but given the number of times we have this in the code base, it could be worth introducing something like this:
```
join(
  step.function_arguments.begin(), step.function_arguments.end(), 
  [](arg) {out << from_expr(ns, step.function, arg);},
  []() {out << ", ";});
```

Reply:
I think you can already do this:
```
auto arg_strings = make_range(step.function_arguments).map<std::string>(
  [&this](const exprt &arg) { return from_expr(ns, step.function, e); });
join_strings(out, arg_strings.begin(), arg_strings.end(), ", ");
```
And if the above works we should likely use it in lots of places :-)

Reply:
The range type doesn't have a difference_type, and thus, that doesn't work.

Separate PR.